### PR TITLE
[Feature][WMS] Add crs information in WMS GetFeatureInfo output when it differs from WGS84

### PR DESCRIPTION
**Pull request by troopa81**
GeoJSON states that CRS should always be OGC:CRS84, but when user requests explicitly a given CRS different from WGS84 on WMS GetFeatureInfo, it's better to expose the output CRS using the old GeoJSON specification [format](http://wiki.geojson.org/GeoJSON_draft_version_6#Specification).

Related to a QWC issue, see [comment](https://github.com/qgis/qwc2/pull/278#issuecomment-1830670920)

**Funded by Eurométropole de Strasbourg**

cc @benoitblanc

### CODE REVIEW

**Review by pathmapper — COMMENTED** [review]
I'm not against this change (it improves the current situation 👍 ), but wonder if it wouldn't be an option to reproject the coordinates to WGS 84 to get RFC 7946 GeoJSON if the request is using a different CRS parameter.

From https://geojson.org/:

> RFC 7946 was published in August 2016 and is the new standard specification of the GeoJSON format, replacing the 2008 GeoJSON specification.

It's now more than 7 years since RFC 7946 and there are some tools already complaining if they are provided with a GeoJSON with CRS member, e.g. https://geojsonlint.com or https://geojson.io/:

image

**Comment by troopa81** [comment]
> I'm not against this change (it improves the current situation 👍 ), but wonder if it wouldn't be an option to reproject the coordinates to WGS 84 to get RFC 7946 GeoJSON if the request is using a different CRS parameter.

You don't want to reproject the coordinates in WGS84 because user requests their GetFeatureInfo result in a given coordinate system (using CRS/SRS parameter) and expect their result in that coordinate system. 

> It's now more than 7 years since RFC 7946 and there are some tools already complaining if they are provided with a GeoJSON with CRS member, e.g. https://geojsonlint.com/ or https://geojson.io/:

Funny thing with https://geojsonlint.com/ is that he just checks if the root GeoJSON object has a "crs" item. But the GeoJSON spectification allows to have any additionnal members... so actually, the GeoJSON IS valid. If I replace "crs" with "srs" everything is fine.

And If I give the linter coordinates in 3857, it doesn't yell. So what its best:
- Having a GeoJSON in a different CRS and not knowing which one ?
- Having a GeoJSON in a different CRS and advertising its CRS with an old, deprecated format (But still GeoJSON valid) ?

I 'd vote the second one. @elpaso Any opinion on this ?

**Comment by elpaso** [comment]
> > I'm not against this change (it improves the current situation 👍 ), but wonder if it wouldn't be an option to reproject the coordinates to WGS 84 to get RFC 7946 GeoJSON if the request is using a different CRS parameter.
> 
> You don't want to reproject the coordinates in WGS84 because user requests their GetFeatureInfo result in a given coordinate system (using CRS/SRS parameter) and expect their result in that coordinate system.
> 
> > It's now more than 7 years since RFC 7946 and there are some tools already complaining if they are provided with a GeoJSON with CRS member, e.g. https://geojsonlint.com/ or https://geojson.io/:
> 
> Funny thing with https://geojsonlint.com/ is that he just checks if the root GeoJSON object has a "crs" item. But the GeoJSON spectification allows to have any additionnal members... so actually, the GeoJSON IS valid. If I replace "crs" with "srs" everything is fine.
> 
> And If I give the linter coordinates in 3857, it doesn't yell. So what its best:
> 
>     * Having a GeoJSON in a different CRS and not knowing which one ?
> 
>     * Having a GeoJSON in a different CRS and advertising its CRS with an old, deprecated format (But still GeoJSON valid) ?
> 
> 
> I 'd vote the second one. @elpaso Any opinion on this ?

Well, I don't have a strong opinion: the first option makes sense because the client should know the response CRS (because it asked for it in the first place) but is also true that the server might not be able to honor the request, explicit is better than implicit and having the crs exposed in the response is not a bad idea.

Would this also affetc WFS3/oapif ?  I implemented CRS in the response even if it wasn't in RFC 7946 because I thought it was too useful to not to, but the CRS is not currently exposed in the response.

**Comment by troopa81** [comment]
> Would this also affetc WFS3/oapif ? I implemented CRS in the response even if it wasn't in RFC 7946 because I thought it was too useful to not to, but the CRS is not currently exposed in the response.

It only affects WMS/GetFeatureInfo for now, but it could (and maybe should?)

**Comment by elpaso** [comment]
> > Would this also affetc WFS3/oapif ? I implemented CRS in the response even if it wasn't in RFC 7946 because I thought it was too useful to not to, but the CRS is not currently exposed in the response.
> 
> It only affects WMS/GetFeatureInfo for now, but it could (and maybe should?)

Yeah, probably better to add it there too.

**Comment by troopa81** [comment]
> Yeah, probably better to add it there too.

Done [f7d5feb](https://github.com/qgis/QGIS/pull/56600/commits/f7d5febf33faa6e63fffcc985487152b99004cd2)

**Comment by troopa81** [comment]
unrelated test failure

File: src/server/services/wfs/qgswfsgetfeature.cpp [hunk 1]
```diff
@@ -1174,6 +1174,21 @@ namespace QgsWfs
...
+        const QString srsName {request.serverParameters().value( QStringLiteral( "SRSNAME" ) )};
+        const QgsCoordinateReferenceSystem destinationCrs { srsName.isEmpty( ) ? QStringLiteral( "EPSG:4326" ) : srsName };
+        if ( ! destinationCrs.isValid() )
+        {
+          throw QgsRequestNotWellFormedException( QStringLiteral( "srsName error: '%1' is not valid." ).arg( srsName ) );
+        }
+
+        json value;
+        QgsJsonUtils::addCrsInfo( value, destinationCrs );
+        for ( auto it : value.items() )
```
Comment (pblottiere) [inline src/server/services/wfs/qgswfsgetfeature.cpp:1187]:
```suggestion
        for ( const auto &it : value.items() )
```

**Review by pblottiere — APPROVED** [review]

**Comment by troopa81** [comment]
> @troopa81 A WFS GetFeature request can also be used to request data in GeoJSON format in a non-EPSG:4326 CRS. As you add toOgcUrn() method, I think the CRS is not added to the GeoJSON provided by WFS GetFeature request. Do you mind to do it to get consistent GeoJSON in QGIS Server ?

I added it in [f7d5feb](https://github.com/qgis/QGIS/pull/56600/commits/f7d5febf33faa6e63fffcc985487152b99004cd2) following @elpaso comment

**Comment by rldhont** [comment]
> > @troopa81 A WFS GetFeature request can also be used to request data in GeoJSON format in a non-EPSG:4326 CRS. As you add toOgcUrn() method, I think the CRS is not added to the GeoJSON provided by WFS GetFeature request. Do you mind to do it to get consistent GeoJSON in QGIS Server ?
> 
> I added it in [f7d5feb](https://github.com/qgis/QGIS/pull/56600/commits/f7d5febf33faa6e63fffcc985487152b99004cd2) following @elpaso comment

Sorry I have not seen it. Thanks !

**Review by rldhont — APPROVED** [review]

**Pull request merged**